### ISSUE TITLE
BytesCache bug leading to duplicate file reads for small files on remote storage

### ISSUE THREAD

**Issue by hayesgb** [intake/filesystem_spec#110]
I'm working on an Azure Datalake Gen2 package using fsspec and encountered strange behavior with BytesCacheing, where the file being read from the datalake and calling compute() on the cached dataframe appears as an exact duplicate of rows of the same dataframe.

It appears to be caused by the _fetch method in BytesCache when reading a file into Dask.  During what should be the first read when calling "dd.read_csv()", the value of start (when none) is passed as zero, but when calling the .compute() method on the dataframe, the file gets read again, leading to the duplication of rows.  Explicitly setting start, end= 0,0 like: 

```
def _fetch(self, start, end):
        # TODO: only set start/end after fetch, in case it fails?
        # is this where retry logic might go?
        if self.start is None and self.end is None:
            # First read
            start, end = 0, 0
            self.cache = self.fetcher(start, end + self.blocksize)
            self.start = start
```

appears to fix it for me.  Have you observed any issues with BytesCacheing elsewhere?

**Comment by hayesgb** [comment]
Continuing to work on the filesystem for Azure Datalake, I looked at the HTTPFileSystem implementation.  There, the method _fetch_all() and AllBytes was was introduced.  I replicated this design pattern in my work on adlfs by adding that same class in my package's core.py file.  But would it make sense to move the AllBytes class from the http.py implementation into core.py?

**Comment by TomAugspurger** [comment]
Just FYI, Martin is out for a few weeks, so responses will be a bit slow here.

At a glance, `AllBytes` looks similar to `core.BytesCache`. I notice that there's some leftover debug code with `blocksize=None` in there.

**Comment by martindurant** [comment]
AllBytes exists for the case where we didn't know the size of the remote HTTP url, because the server wouldn't say, and then we ended up downloading all the data, which is less than the prescribed block size. This is a little specific; in the case we know the size, and it is smaller than the block size, BytesCache should behave the same way. I suppose it is a reasonable caching policy anyway, by which we say "download everything" regardless of block size - this is done in testing in places. It is *not* a good general-purpose cache, because we expect to want to read parts of very large files in a random-access manner!

**Comment by martindurant** [comment]
(it would be better to isolate the bug in BytesCache and test against one of the backend implementations which use it)

**Comment by cottrell** [comment]
In case this is useful ... we just hit this to through a dask dataframe read_csv call. 

dd.read_csv('s3....').compute()

The file gets doubled -1 (column appears as a row twice).

Reverting to fsspec 0.4.1 seems to fix ... I haven't chased this all the way down into the fsspec stack.

**Comment by cottrell** [comment]
> @cottrell what version of fsspec did you observe the buggy behavior on?
> 
> > The file gets doubled -1 (column appears as a row twice).
> 
> Can you clarify that?

Version 0.4.2 was buggy. 0.4.1 was ok. 

This is a minimal version dump of a test env I used:

```
$ pip list
Package           Version
----------------- ---------
astroid           2.2.5
backcall          0.1.0
boto3             1.9.216
botocore          1.12.216
certifi           2019.6.16
cloudpickle       1.2.1
dask              2.3.0
decorator         4.4.0
docutils          0.15.2
fsspec            0.4.2
ipython           7.7.0
ipython-genutils  0.2.0
isort             4.3.21
jedi              0.15.1
jmespath          0.9.4
lazy-object-proxy 1.4.1
numpy             1.17.1
pandas            0.25.1
parso             0.5.1
pexpect           4.7.0
pickleshare       0.7.5
pip               19.2.2
prompt-toolkit    2.0.9
ptyprocess        0.6.0
Pygments          2.4.2
pylint            2.3.1
python-dateutil   2.8.0
pytz              2019.2
s3fs              0.3.3
s3transfer        0.2.1
setuptools        41.0.1
six               1.12.0
toolz             0.10.0
traitlets         4.3.2
urllib3           1.25.3
wcwidth           0.1.7
wheel             0.33.4
```

**Comment by cottrell** [comment]
> @cottrell , it would help us a lot if you could post a code sample which, in isolation, shows the problem - something like Tom's code above (which does _not_ seem to reproduce it).

Here is some benign data that reproduces the error with the versions above:

```
import dask.dataframe as dd
import pandas as pd
df = pd.DataFrame(
    {'TICKER': {0: 'A', 1: 'AA', 2: 'AAAP', 3: 'AAC', 4: 'AAL'},
         'MIN_DATE': {0: '1999-11-18',
               1: '2016-11-01',
                 2: '2015-11-11',
                   3: '2014-10-02',
                     4: '2005-09-27'},
          'MAX_DATE': {0: '2019-08-26',
                1: '2019-08-26',
                  2: '2018-02-09',
                    3: '2019-08-26',
                      4: '2019-08-26'},
           'COUNT': {0: '4973', 1: '708', 2: '566', 3: '1233', 4: '3502'}})


filename = 's3://SOMELOCATION/test.csv'
df.to_csv(filename)
dd.read_csv(filename, dtype=str).compute() # ok
dd.read_csv(filename).compute() # fails
```

It is the count column that complains.


```
ValueError: Mismatched dtypes found in `pd.read_csv`/`pd.read_table`.

+------------+---------+----------+
| Column     | Found   | Expected |
+------------+---------+----------+
| COUNT      | object  | int64    |
| Unnamed: 0 | float64 | int64    |
+------------+---------+----------+

The following columns also raised exceptions on conversion:

- COUNT
  ValueError("invalid literal for int() with base 10: 'COUNT'")

Usually this is due to dask's dtype inference failing, and
*may* be fixed by specifying dtypes manually by adding:

dtype={'COUNT': 'object',
       'Unnamed: 0': 'float64'}

to the call to `read_csv`/`read_table`.
```


But you can see on debug that the number of rows is doubled:

```
In [23]: %debug
*** NameError: name 'pdb' is not defined
> /home/ubuntu/anaconda3/envs/tt/lib/python3.7/site-packages/dask/dataframe/io/csv.py(178)coerce_dtypes()
    176             rule.join(filter(None, [dtype_msg, date_msg]))
    177         )
--> 178         raise ValueError(msg)
    179
    180

ipdb> up
> /home/ubuntu/anaconda3/envs/tt/lib/python3.7/site-packages/dask/dataframe/io/csv.py(74)pandas_read_text()
     72     df = reader(bio, **kwargs)
     73     if dtypes:
---> 74         coerce_dtypes(df, dtypes)
     75
     76     if enforce and columns and (list(df.columns) != list(columns)):

ipdb> df.shape
(11, 5)

ipdb> df
    Unnamed: 0  TICKER    MIN_DATE    MAX_DATE  COUNT
0          0.0       A  1999-11-18  2019-08-26   4973
1          1.0      AA  2016-11-01  2019-08-26    708
2          2.0    AAAP  2015-11-11  2018-02-09    566
3          3.0     AAC  2014-10-02  2019-08-26   1233
4          4.0     AAL  2005-09-27  2019-08-26   3502
5          NaN  TICKER    MIN_DATE    MAX_DATE  COUNT
6          0.0       A  1999-11-18  2019-08-26   4973
7          1.0      AA  2016-11-01  2019-08-26    708
8          2.0    AAAP  2015-11-11  2018-02-09    566
9          3.0     AAC  2014-10-02  2019-08-26   1233
10         4.0     AAL  2005-09-27  2019-08-26   3502
```

**Comment by cottrell** [comment]
Yes, it appears to be fixed with 0.4.3.
